pkg/controllers/management/cloudcredential/controller.go

```diff
@@ -2,32 +2,180 @@ package cloudcredential
 	"k8s.io/apimachinery/pkg/runtime"
 )
 
+const (
+	harvesterCloudCredentialTokenChecksumAnnotation = "management.cattle.io/harvester-token-checksum"
+	harvesterCloudCredentialFinalizer               = "management.cattle.io/harvester-token-cleanup"
+)
+
 type Controller struct {
 	managementContext *config.ManagementContext
+	secretClient      corecontrollers.SecretClient
+	tokenClient       mgmtcontrollers.TokenClient
+	apply             apply.Apply
 }
 
-func Register(ctx context.Context, management *config.ManagementContext) {
+func Register(ctx context.Context, management *config.ManagementContext, clients *wrangler.Context) {
 	m := Controller{
 		managementContext: management,
+		secretClient:      clients.Core.Secret(),
+		tokenClient:       clients.Mgmt.Token(),
+		apply: clients.Apply.WithCacheTypes(
+			clients.Core.Secret(),
+			clients.Mgmt.Token(),
+		),
 	}
 	management.Core.Secrets("").AddHandler(ctx, "management-cloudcredential-controller", m.ccSync)
+	if features.Harvester.Enabled() {
+		clients.Core.Secret().OnChange(ctx, "harvester-cloud-credential-token", m.syncHarvesterToken)
+	}
+}
+
+// syncHarvesterToken will extend the ttl of a token owned by a harvester cloud credential to infinite, as well as
+// handle deletion when the cloud credential secret is deleted.
+func (c *Controller) syncHarvesterToken(key string, secret *v1.Secret) (*v1.Secret, error) {
+	if secret == nil || secret.Namespace != namespace.GlobalNamespace {
+		return secret, nil
+	}
+
+	if len(secret.Data["harvestercredentialConfig-kubeconfigContent"]) == 0 {
+		// not a (valid) harvester cloud credential
+		return secret, nil
+	}
+
+	var err error
+
+	d := sha256.Sum256(secret.Data["harvestercredentialConfig-kubeconfigContent"])
+	checksum := hex.EncodeToString(d[:])
+
+	if secret.DeletionTimestamp == nil && secret.Annotations[harvesterCloudCredentialTokenChecksumAnnotation] == checksum {
+		// token TTL has already been extended and tokens have not been rotated
+		return secret, nil
+	} else if secret.DeletionTimestamp != nil && !slices.Contains(secret.Finalizers, harvesterCloudCredentialFinalizer) {
+		// token TTL was never extended
+		return secret, nil
+	}
+
+	objs := make([]runtime.Object, 0)
+	if secret.DeletionTimestamp == nil {
+		if !slices.Contains(secret.Finalizers, harvesterCloudCredentialFinalizer) {
+			secret = secret.DeepCopy()
+			secret.Finalizers = append(secret.Finalizers, harvesterCloudCredentialFinalizer)
+			secret, err = c.secretClient.Update(secret)
+			if err != nil {
+				return nil, fmt.Errorf("unable to update harvester cloud credential secret %s: %w", key, err)
+			}
+		}
+
+		// in practice a kubeconfig will only ever have one token, but we need to handle the case where users may be
+		// modifying the secret directly and properly extend/delete tokens as necessary.
+		tokenNames, err := c.tokensFromHarvesterCloudCredential(secret)
+		if err != nil {
+			return nil, err
+		}
+
+		for tokenName := range tokenNames {
```

Comment:
Just for the reviewer, I tested this behavior with the generated kubeconfig, the token's ttl was set to 0. I then created a new kubeconfig via download kubeconfig and added a new user to the kubeconfig, also had the ttl set to 0. I then removed the original user, and it deleted the associated token (as expected). This is never really expected during normal operation, but I wanted to make sure that the edge cases were covered since users may have done some interesting things to their tokens. This also has the benefit of not extending any user created tokens, but still computing the checksum so the controller doesn't run all the time.